module/vision/BallDetector/src/BallDetector.cpp

```diff
@@ -106,137 +108,167 @@ namespace module::vision {
 
-                // We still emit if there are no balls otherwise other modules are not aware that we can no longer see a
-                // ball, and they will keep using the last message sent. See head behaviour, where we use 'with' balls
-                // and react based on that last balls message.
+                // Balls is still emitted even if there are no balls, to indicate to other modules that no balls are
+                // currently visible
                 if (clusters.empty()) {
                     log<NUClear::DEBUG>("Found no balls.");
                     emit(std::move(balls));
                     return;
                 }
 
+                // Reserve the memory in advance for efficiency
                 balls->balls.reserve(clusters.size());
 
-                balls->id        = horizon.id;
-                balls->timestamp = horizon.timestamp;
-                balls->Hcw       = horizon.Hcw;
+                balls->id        = horizon.id;         // camera id
+                balls->timestamp = horizon.timestamp;  // time when the image was taken
+                balls->Hcw       = horizon.Hcw;        // world to camera transform at the time the image was taken
+
+                // World to camera transform, to be used in for loop below
+                const Eigen::Affine3f Hcw(horizon.Hcw.cast<float>());
 
+                // CHECK EACH CLUSTER FOR VALID BALL
                 for (auto& cluster : clusters) {
                     Ball b;
+                    b.covariance = config.ball_angular_cov.asDiagonal();
 
-                    // Average the cluster to get the cones axis
-                    Eigen::Vector3f axis = Eigen::Vector3f::Zero();
+                    // FIND CENTRAL AXIS OF BALL
+                    // Add up all the unit vectors of each point (camera to point in world space) in the cluster to find
+                    // an average vector, which represents the central cone axis
+                    // uBCw: unit vector from camera to ball central axis in world space
+                    Eigen::Vector3f uBCw = Eigen::Vector3f::Zero();
                     for (const auto& idx : cluster) {
-                        axis += rays.col(idx);
+                        uBCw += uPCw.col(idx);
                     }
-                    axis /= cluster.size();
-                    axis.normalize();
+                    uBCw.normalize();  // get cone axis as a unit vector
 
-                    // Find the ray with the greatest distance from the axis
-                    // Should we use the average distance instead?
+                    // FIND UNIT RADIUS OF BALL
+                    // Find the ray (uPCw) with the greatest distance from the central axis (uBCw) to then determine the
+                    // largest radius possible from the edge points available. The radius is the upper bound on the ball
+                    // radius if it were a metre away. arccos(radius) is the angle between the central ball
```

Comment:
I dont think the ball being a meter away has anything to do with it. This is just an upper bound on the ball radius

Reply:
How is it an upper bound on the ball radius? If it's a unit vector, then projecting it out further than a unit (one metre) would give a larger radius? It has to be an upper bound from a particular perspective, since we know in reality the radius is `field.ball_radius`?

Reply:
Because its not actually a radius that we are measuring, its (`cos` of) the angle between the cone axis and the ball edge. Its an upper bound because we are looking for the largest such angle. We could just as feasibly take the minimum angle (the lower bound) or the mean value. Literally the next thing we do is use the angular distance formula to figure out what distance the ball has to be away from the camera for the "radius" we calculate to make sense because, as you said, the actual radius is fixed to the radius of the ball.

Reply:
I think at one point we called this the `gradient` of the cone (I think it was in the message we called it this). Which is also technically correct, but the quantity we are actually measuring isn't actually the gradient of the cone either.

Reply:
The naming of it seems not great to me. Angular radius seems like a better name.

Reply:
Its also a longer name and more annoying to type :stuck_out_tongue: just comment it lol